File: 프로그래머스/2/131127. 할인 행사/할인 행사.js
function solution(want, number, discount) {
    var answer = 0;
    const SALE_PERIOD = 10;
    let wantList = [];
    
    for (let i = 0; i < want.length; i++) {
        let count = number[i];
        while (count > 0) {
            wantList.push(want[i]);
            count--;
        }
    }
    
    for (let start = 0; start < discount.length - SALE_PERIOD + 1; start++) {
        let wantListCopy = Array.from(wantList);
        
        for (let idx = start; idx < start + SALE_PERIOD; idx++) {
            if (wantListCopy.includes(discount[idx])) {
                let firstIndex = wantListCopy.findIndex(x => x == discount[idx]);
                wantListCopy.splice(firstIndex, 1);
            }
        }
        answer += wantListCopy.length == 0? 1: 0;
    }
    
    return answer;
}

File: 프로그래머스/2/131127. 할인 행사/할인 행사.java
import java.util.*;

class Solution {
    private static final int SALE_PERIOD = 10;
    
    public int solution(String[] want, int[] number, String[] discount) {
        int answer = 0;
        
        final Map<String, Integer> wantMap = new HashMap<>();
        for (int i = 0; i < want.length; i++) {
            wantMap.put(want[i], number[i]);
        }
        
        Map<String, Integer> windowMap = new HashMap<>();
        for (int i = 0; i < SALE_PERIOD; i++) {
            windowMap.put(discount[i], windowMap.getOrDefault(discount[i], 0) + 1);
        }
        
        if (matches(wantMap, windowMap)) {
            answer++;
        }
        
        for (int i = SALE_PERIOD; i < discount.length; i++) {       
            windowMap.put(discount[i - SALE_PERIOD], windowMap.getOrDefault(discount[i - SALE_PERIOD], 0) - 1);
            if (windowMap.get(discount[i - SALE_PERIOD]) <= 0) {
                windowMap.remove(discount[i - SALE_PERIOD]);
            }
            
            windowMap.put(discount[i], windowMap.getOrDefault(discount[i], 0) + 1);
            
            if (matches(wantMap, windowMap)) {
                answer++;
            }
        }
        
        
        return answer;
    }
    
    private boolean matches(Map<String, Integer> wantMap, Map<String, Integer> windowMap) {
        for (Map.Entry<String, Integer> entry: wantMap.entrySet()) {
            String key = entry.getKey();
            Integer value = entry.getValue();
            
            if (windowMap.getOrDefault(key, 0) < value) {
                return false;
            }
        }
        return true;
    }
}